### PR TITLE
Optimise List.map by chunking the intermediate representation.

### PR DESCRIPTION
**Pull request by paurkedal**
Here is a version of the "tupled" map optimisation which also keeps the current handing of the prefix, but if you also want to simplify the code, I can drop that, at the expense of slightly reduced performance.

Compare "base" with "tuple" here to see the difference between keeping the prefix handing or not:
image
The current base performance can be seen in https://discuss.ocaml.org/t/a-new-list-map-that-is-both-stack-safe-and-fast/865/32.

### CODE REVIEW

**Comment by aalekseyev** [comment]
What happened to "base" performance between this plot and the one at [the link](https://discuss.ocaml.org/t/a-new-list-map-that-is-both-stack-safe-and-fast/865/32)? It looks way better here. In fact here it's better than "tupled" on this plot so why would we want "tupled" then? :)

**Comment by aalekseyev** [comment]
Ah, sorry after reading your comment I understand. On this plot "base" is base with this PR applied.

**Comment by paurkedal** [comment]
Yes, that was with the patch applied, though I copied it out of base, since dependencies for the testing framework failed to compile with the master of base (tried opam pin and jbuilder).

**Comment by seliopou** [comment]
I had the same reaction as @aalekseyev when I saw the chart. Given that these sorts of changes also have to go through internal review, could you generate a chart with labels that make it clear what we should be comparing? Something like `base-old` for the current mater, and `base-new` or `base-tupled` for your change?

**Comment by aantron** [comment]
cc @bluddy

**Comment by bluddy** [comment]
Any idea how much the chunk size matters? Would be nice to just graph all chunk sizes up to... 12?

**Comment by paurkedal** [comment]
@seliopou I'll update the graph when it's finished.

@bluddy I tried 12 and 7. 12 was slower than 8 in fact.  7 was also slower than 8, but that seemed like what one would naively expect.

**Comment by bluddy** [comment]
@paurkedal As discussed on discuss.ocam.org, the chunk size is probably sensitive to the architecture, but I see nothing wrong with being biased towards x86-64. Perhaps you want to flesh it out some more with 9, and just make sure you've found an optimal point?

**Comment by paurkedal** [comment]
@bluddy Will do. If you don't mind, I think it'll be most instructive without the the base prefix special case, so I'll use the original tupled.

**Comment by aantron** [comment]
FYI For my implementation, I first did some kind of near-"binary search" between sizes 2 and 64, then tried every size between 4 and 16 or so. You can probably skip the "binary search" part, but I suggest trying a bunch of different chunk sizes around 8.

**Comment by paurkedal** [comment]
@aantron Yeah, I'm currently crushing a new graph for 7, 8, 9, 10, 11.

**Comment by paurkedal** [comment]
I extended to 4-12 and the result is:
image
The graph supports increasing chunks to 9, at least.  12 looks better than I though, but only for longer lists.

**Comment by bluddy** [comment]
Great job. I agree that 9 looks pretty good. If we could switch from direct to tupled at 100,000, 12 would dominate, but I don't think that's safe to do given the stack size.

It looks like tupled-9 even beats the direct strategy on small lists. Is that right?

**Comment by seliopou** [comment]
@paurkedal can you please include the current implementation of map as either the baseline or an additional series in the chart? It's good that the implementation improves over the naive implementation but the comparison that's important here is between the current Base implementation and the change you're proposing.

**Comment by yminsky** [comment]
A few thoughts:

- First, this looks like good work, and thanks for taking the time to do it.
- This all seems like a good argument for tail-recursion modulo constructors. The fact that all this manual tweakery helps is pretty sad.
- While nice, the practical impact of this seems limited, since performance for very long lists is just not that important. There's a reason the initial complex implementation of map didn't optimize the long-list case, which is that, while it's important that it doesn't blow up, code that relies on the performance of mapping over very long lists is broken in other ways.

All of which is to say: we should seriously consider whether the added complexity is worth the improvement gained here. I weakly lean in favor of merging, but the whole thing makes me a little sad.

**Comment by yminsky** [comment]
Also, agreed with @seliopou that it would be nice to see the improvement measured against the current implementation in Base.

**Comment by bluddy** [comment]
> This all seems like a good argument for tail-recursion modulo constructors. The fact that all this manual tweakery helps is pretty sad.

Perhaps, but TRMC adds mutation that won't be tolerated well (at least initially) by multicore. I think we can only consider it after multicore is already merged, so we're looking at ~2022  as a target to possibly get TRMC.

> While nice, the practical impact of this seems limited, since performance for very long lists is just not that important.

Another perspective is that lists are the default data structure of OCaml, and we need to do what we can to make them as performant as possible without cheating (aka the Batteries approach). Library code can be made complex to achieve better performance precisely because it's hidden from the user.

BTW One could envision an eventual Flambda pass that, after unrolling loops, looks for several list allocations that will definitely be destructed (similar to tuple elimination), and converts them to a single allocation.

**Comment by paurkedal** [comment]
Here is a plot using base a reference with x-range extended down to 10:
image
@seliopou, @yminsky The new code is slightly simpler than the current, but if small list are the priority, we'd clearly need to keep the current code, as well.  So, I understand if you want to keep the code lean.

@bluddy If by direct strategy, you mean stdlib then the result for 10 elements is 35.20ns for tupled vs 35.58ns for stdlib, assuming the tests are representative for such small lists.

Thanks for the hint about TRMC, that's would be a much nicer solution.  (Naively I don't see how it would conflict with multicore, since I'd expect any intervening concurrency primitives to break the conditions for the optimisation, and the resulting list not to be passed on to the caller before it's frozen.)

**Comment by yminsky** [comment]
Yeah, small lists should definitely be the priority, so I don't think it's worth considering a patch that worsens the numbers there. Is there a version of your patch that strictly improves upon the current implementation?

**Comment by paurkedal** [comment]
@yminsky The current version of the PR combines `count_map` from base with tupled-8.  So it's theoretically the same for < 1000 unless the compiler does something quite unexpected. The comparison can be seen in the first plot, where "base" and "base-new" seems to coincide within error margins at 1000.  The other two plots just the tupled code.  If you wish to merge, I suggest I replace the > 1000 fallback with tupled-9 first.

**Comment by yminsky** [comment]
Yup, that sounds like the right course of action.

**Comment by paurkedal** [comment]
I pushed the proposed change, but before you merge: How do I run the testsuite?

A final performance comparison:
image

**Comment by yminsky** [comment]
Thanks! We're pushing on getting this through our internal review process now.

File: src/list.ml [hunk 1]
```diff
@@ -309,7 +309,27 @@ let rec count_append l1 l2 count =
 
 let append l1 l2 = count_append l1 l2 0
 
-let map_slow l ~f = rev (rev_map ~f l)
+let tail_map xs ~f =
```
[the remark this answers is not in the document]
Comment (paurkedal) [reply src/list.ml]: Something like this?:
```ocaml
(* A pure list map using plain tail recursion must store an intermediate
 * (reversed) representation.  The following implementation reduce the
 * performance and allocation impact by chucking 9 original elements per
 * intermediate list node. *)
```

**Comment by seliopou** [comment]
This will appear in our next push to GitHub. Thanks for your contribution!

**Comment by seliopou** [comment]
Folks, I have some bad news. This change breaks our JavaScript builds. One of our goals with base is to provide a cross-platform standard library alternative, so unfortunately I'm going to have to revert the change until the breakage is resolved.

The specific issue we're coming across is stack overflows when we compile our tests to JavaScript and run them on node.js. This isn't limited to Base tests, but also shows up when running tests for [bin_prot](https://github.com/janestreet/bin_prot).

We added the following simple tests to base's test suite:

```diff
diff --git a/lib/base/test/test_list.ml b/lib/base/test/test_list.ml                                                                                                                                                                           
--- a/lib/base/test/test_list.ml                                                                                                                                                                                                               
+++ b/lib/base/test/test_list.ml                                                                                                                                                                                                               
@@ -169,6 +169,10 @@                                                                                                                                                                                                                           
     let v = lazy (range 1 100_000) in                                                                                                                                                                                                         
     fun () -> Lazy.force v                                                                                                                                                                                                                    
                                                                                                                                                                                                                                               
+  let long2 =                                                                                                                                                                                                                                 
+    let v = lazy (range 2 100_001) in                                                                                                                                                                                                         
+    fun () -> Lazy.force v                                                                                                                                                                                                                    
+                                                                                                                                                                                                                                              
   let l1 = [1;2;3;4;5;6;7;8;9;10]                                                                                                                                                                                                             
 end                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                               
@@ -187,6 +191,21 @@                                                                                                                                                                                                                           
   ignore (rev_append long long:int list);                                                                                                                                                                                                     
   true                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                               
+let%test _ =                                                                                                                                                                                                                                  
+  let long1 = Test_values.long1 () in                                                                                                                                                                                                         
+  let long2 = Test_values.long2 () in                                                                                                                                                                                                         
+  (map long1 ~f:(fun x -> x + 1)) = long2                                                                                                                                                                                                     
+                                                                                                                                                                                                                                              
+let test_ordering n =                                                                                                                                                                                                                         
+  let l = List.range 0 n in                                                                                                                                                                                                                   
+  let r = ref [] in                                                                                                                                                                                                                           
+  let _ : unit list = List.map l ~f:(fun x -> r := x :: !r) in                                                                                                                                                                                
+  [%test_eq: int list] l (List.rev !r)                                                                                                                                                                                                        
+                                                                                                                                                                                                                                              
+let%test_unit _ = test_ordering 10                                                                                                                                                                                                            
+let%test_unit _ = test_ordering 1000                                                                                                                                                                                                          
+let%test_unit _ = test_ordering 1_000_000                                                                                                                                                                                                     
+                                                                                                                                                                                                                                              
 let%test _ = for_all2_exn [] [] ~f:(fun _ _ -> assert false)                                                                                                                                                                                  
                                                                                                                                                                                                                                               
 let%test _ = find_mapi [0;5;2;1;4] ~f:(fun i x -> if i = x then Some (i+x) else None) = Some 0
```

The first and and fourth of the new tests break when run against the new implementation of `List.map` under JavaScript. We reverted just the change to `List.map` and ran the new tests against the original implementation and everything passed. This also fixed the other failures we were seeing across our tree.

Several of us involved in maintaining Base and the JavaScript build have tried looking into this without much success. Unfortunately we can't leave our internal build in this state much longer while we investigate, so I'll unfortunately have to ask you all to conduct your own independent investigation into the cause of these stack overflows. Once the issue is identified and fixed (wherever it may lie), we'll be happy to reconsider this contribution for merging. If we can provide any additional information, don't hesitate to ask.

**Comment by paurkedal** [comment]
Sure, I'll look into it.  Though I'm struggling with running the inline tests.  After finding https://github.com/janestreet/ppx_inline_test/issues/8 I created a subdirectory with
```
(executable
 ((name test_runner)
  (modules (test_runner))
  (libraries (base_test ppx_inline_test.runner.lib))))

(alias
 ((name    runtest)
  (deps    (test_runner.exe))
  (action  (run ${<} inline-test-runner base_test))))
```
and
```ocaml
test_runner.ml
```
The resulting executable can be run with `test_runner.exe inline-test-runner base` seems to invoke the right entry point, but does not run any tests. I also tried with `base_test` as argument.  Does jenga automatically pass `-inline-test-lib` while jbuilder does not?  Or, any other idea what I'm missing?

(Once I can run the `.exe` I assume `js_of_ocaml +nat.js +weak.js src/runtime.js test_runner.bc` gives the program to test under nodejs.)

**Comment by bluddy** [comment]
This probably has to do with js_of_ocaml's weak heuristics for tail calls. I'm guessing that calling rise from dive makes jsoo miss the tail calls. To simplify , dive could return a pair of the list of bs and the list of processed xs. This would be followed by a separate call to rise. Hopefully this will be enough for jsoo's optimizer to identify the tail calls and turn them into loops.

**Comment by hhugo** [comment]
Js_of_ocaml does not miss tail calls in that case.
It seams that  the hardcoded limit of 1000 non tail call is too high in a JavaScript context.  

Bellow is the generated JavaScript for `count_map` 
```
function count_map(f,l,ctr)
     {if(l)
       {var _zP_=l[2],_zQ_=l[1];
        if(_zP_)
         {var _zR_=_zP_[2],_zS_=_zP_[1];
          if(_zR_)
           {var _zT_=_zR_[2],_zU_=_zR_[1];
            if(_zT_)
             {var _zV_=_zT_[2],_zW_=_zT_[1];
              if(_zV_)
               {var
                 tl=_zV_[2],
                 x5$0=_zV_[1],
                 f1=caml_call1(f,_zQ_),
                 f2=caml_call1(f,_zS_),
                 f3=caml_call1(f,_zU_),
                 f4=caml_call1(f,_zW_),
                 f5=caml_call1(f,x5$0);
                if(caml_call2(_s_,ctr,1e3))
                 {var bs=0,xs=tl;
                  for(;;)
                   {if(xs)
                     {var _zH_=xs[2];
                      if(_zH_)
                       {var _zI_=_zH_[2];
                        if(_zI_)
                         {var _zJ_=_zI_[2];
                          if(_zJ_)
                           {var _zK_=_zJ_[2];
                            if(_zK_)
                             {var _zL_=_zK_[2];
                              if(_zL_)
                               {var _zM_=_zL_[2];
                                if(_zM_)
                                 {var _zN_=_zM_[2];
                                  if(_zN_)
                                   {var _zO_=_zN_[2];
                                    if(_zO_)
                                     {var
                                       xs$0=_zO_[2],
                                       x8=_zO_[1],
                                       x7=_zN_[1],
                                       x6=_zM_[1],
                                       x5=_zL_[1],
                                       x4=_zK_[1],
                                       x3=_zJ_[1],
                                       x2=_zI_[1],
                                       x1=_zH_[1],
                                       x0=xs[1],
                                       y0$0=caml_call1(f,x0),
                                       y1$0=caml_call1(f,x1),
                                       y2$0=caml_call1(f,x2),
                                       y3$0=caml_call1(f,x3),
                                       y4$0=caml_call1(f,x4),
                                       y5$0=caml_call1(f,x5),
                                       y6$0=caml_call1(f,x6),
                                       y7$0=caml_call1(f,x7),
                                       y8$0=caml_call1(f,x8),
                                       bs$0=[0,[0,y0$0,y1$0,y2$0,y3$0,y4$0,y5$0,y6$0,y7$0,y8$0],bs],
                                       bs=bs$0,
                                       xs=xs$0;
                                      continue}}}}}}}}}
                    var ys$1=nontail_map(xs,f),ys=ys$1,param=bs;
                    for(;;)
                     {if(param)
                       {var
                         param$0=param[2],
                         match=param[1],
                         y8=match[9],
                         y7=match[8],
                         y6=match[7],
                         y5=match[6],
                         y4=match[5],
                         y3=match[4],
                         y2=match[3],
                         y1=match[2],
                         y0=match[1],
                         ys$0=
                          [0,
                           y0,
                           [0,y1,[0,y2,[0,y3,[0,y4,[0,y5,[0,y6,[0,y7,[0,y8,ys]]]]]]]]],
                         ys=ys$0,
                         param=param$0;
                        continue}
                      var _zX_=ys;
                      break}
                    break}}
                else
                 var _zX_=count_map(f,tl,ctr + 1 | 0);
                return [0,f1,[0,f2,[0,f3,[0,f4,[0,f5,_zX_]]]]]}
              var
               f1$0=caml_call1(f,_zQ_),
               f2$0=caml_call1(f,_zS_),
               f3$0=caml_call1(f,_zU_),
               f4$0=caml_call1(f,_zW_);
              return [0,f1$0,[0,f2$0,[0,f3$0,[0,f4$0,0]]]]}
            var
             f1$1=caml_call1(f,_zQ_),
             f2$1=caml_call1(f,_zS_),
             f3$1=caml_call1(f,_zU_);
            return [0,f1$1,[0,f2$1,[0,f3$1,0]]]}
          var f1$2=caml_call1(f,_zQ_),f2$2=caml_call1(f,_zS_);
          return [0,f1$2,[0,f2$2,0]]}
        var f1$3=caml_call1(f,_zQ_);
        return [0,f1$3,0]}
      return 0}
```

**Comment by paurkedal** [comment]
I cannot reproduce the issue with OCaml-4.04.2 and js_of_ocaml-2.8.4 or OCaml-4.05.0 and js_of_ocaml-3.0 run under nodejs v4.7.2.  My test setup is otherwise a mix of opam packages and checkouts of master branches and latest version tags of dependencies.  My ulimit -s is 131072, but reducing it to 1024 did not trigger a stack overflow, so nodejs probably has its own stack limit.

So, also considering @hhugo analysis, the stack overflow is probably caused by the addition of a constant added on top of the 1000 frames, that tips it over the limit.

You could try to reduce the number of frame to 900 or 800, and I'd make a new benchmark.  If I should be of more help, you'd need to document your test setup in sufficient detail that I can reproduce the issue.  Or we could decide the performance gain is not worth the effort.

**Comment by hhugo** [comment]
I wonder if we should compute/adjust the limit at runtime.
The following code produce very different result depending on how you run it.
```ocaml
let stack_size =
  let c = ref 0 in
  let rec f () =
    incr c; 
    () :: f ()
  in
  try
    ignore (f ());
    assert false
  with Stack_overflow -> !c
```
nodejs: 15710
native: 523544
toplevel in the browser: 10448
toplevel bytecode: 262057

**Comment by hhugo** [comment]
I think we should apply the same strategy that was used in the stdlib and have a much smaller threshold when using non standard backend.

https://github.com/ocaml/ocaml/blob/trunk/stdlib/list.ml#L72

```ocaml
let threshold =
  match Sys.backend_type with
  | Sys.Native | Sys.Bytecode -> 1000
  (* We don't know the size of the stack, better be safe and assume it's
     small. *)
  | Sys.Other _ -> 50
```

I'll resurrect this feature and make this change.

**Comment by xclerc** [comment]
Merged as https://github.com/janestreet/base/commit/2c926a288a3a013d269f747efc8938d12c26b1a2.

**Comment by hhugo** [comment]
Thanks @paurkedal for your contribution.

**Pull request closed without merging**